### ISSUE TITLE
[Feature] Replace pyUSB with pyhidapi - Install User-Level-Service

### ISSUE THREAD

**Issue by suabo** [JSubelj/g910-gkey-macro-support#73]
I tried adding a .rules file for the supported devices to run the program on user level and could get information from keyboard. But as soon as you want to detach the kernel module you get a permission denied. Without detaching it is not possible to read or write to the USB. 

The only way to allow this would be to replace pyusb with [pyhidapi](https://github.com/apmorton/pyhidapi).

This would be the next big upgrade to the driver and would fix #53, since it is related to detaching the kernel driver via pyusb.

I will update this post if I am gonna work on it.

**Comment by suabo** [comment]
I tested a bit more on running on user level, since there is not only the usb we need to access. There is also the uinput device we need to create in order to send keyboard signals. Because of that I added the following to the rules file:  
`KERNEL=="uinput", TAG+="uaccess"`

This will allow for logged in user to access uinput. I needed to run following commands to reload the udev rules after adding them to /etc/udev/rules.d/60-g910-gkeys.rules to make it work:  
`udevadm control --reload-rules && udevadm trigger`

After this I could run `python3 cli_entry_point.py -l create` with user privileges(see #68). This is because we only need access to uinput here to create the mapping.

Just wanted to document this knowledge somewhere. No updates on implementing pyhidapi yet.

**Comment by braoult** [comment]
> This will allow for logged in user to access uinput.

I am unsure what *logged-in user* means here (vs *currently active*). For example, we can have two users (let say `Alice` and `Bob`) **both having their own X server and session**, **both active** (we can just switch user with `Control-alt-F7` and `control-alt-F8`, all Alice processes continue to work even if we are on Bob session).

How will uinput work in this case ? I mean, I am `Alice on vt7`, all keyboard and mouse events come to my session. I switch to `Bob on vt8`, all keyboard and mouse events come to my session (`Alice` and `Bob` share the same `seat`, in X terminology, seat meaning . For example, in my situation :
``` console
Alice@lorien:~$ pstree -cal
-lightdm
   |-Xorg -core :0 -seat seat0 -auth /var/run/lightdm/root/:0 -nolisten tcp vt7 -novtswitch
   |   `-{Xorg}
   |-Xorg -core :1 -seat seat0 -auth /var/run/lightdm/root/:1 -nolisten tcp vt8 -novtswitch
   |   `-{Xorg}
   |-lightdm --session-child 14 21
   |   |-xfce4-session
   |   |   |- [ Alice session programs ]
   |   |   |- ...
   |   |-{lightdm}
   |   `-{lightdm}
   |-lightdm --session-child 16 30
   |   |-xfce4-session
   |   |   |- [ Bob session programs ]
   |   |   |- ...
   |   |-{lightdm}
   |   `-{lightdm}
   |-{lightdm}
   `-{lightdm}
```
The thing is that we should not mess with permissions, and have the correct keyboard input sent to the correct X server.

I made the test: With [my configuration](https://github.com/JSubelj/g910-gkey-macro-support/blob/master/docs/Configuration-strict-user-environment.org), `xev` shows the correct mapping (e.g. pressing `G3` shows `F15` `KeyPress` and `KeyRelease` events) on both `Alice/vt7` and `Bob/vt8`. It means the two X servers are able to grab/release the devices when they become `active/inactive`.

I wonder what `KERNEL=="uinput", TAG+="uaccess"` means exactly.  I believe it could be fine, or the whole `uinput` would be totally wrong on an `Unix-like` (the `multi-user` part of it) system like `GNU/Linux`.

See [This link](https://www.freedesktop.org/wiki/Software/systemd/multiseat/#definitionofterms) to understand what is a `seat` (display, input devices, audio, camera, etc...), `session`, and `multi-session` (multiple sessions on the same seat).

**Comment by suabo** [comment]
Maybe the topic is a not clear enough what the issue is really about.
In the first place I tried to make it runable without root permissions, which can't be a bad thing.

I also made some progress on testing what's the gkey mode all about. With the current rules applied to udev you can start the program with user privileges without permission problems (even detaching an re-attaching the kernel driver). I think this is a big step into the right direction. 

This is my current rules file:
```
SUBSYSTEMS=="usb", ATTRS{idVendor}=="046d", ATTRS{idProduct}=="c335", TAG+="uaccess"
SUBSYSTEMS=="usb", ATTRS{idVendor}=="046d", ATTRS{idProduct}=="c33f", TAG+="uaccess"
SUBSYSTEMS=="usb", ATTRS{idVendor}=="046d", ATTRS{idProduct}=="c33e", TAG+="uaccess"
SUBSYSTEMS=="usb", ATTRS{idVendor}=="046d", ATTRS{idProduct}=="c32b", TAG+="uaccess"
SUBSYSTEMS=="usb", ATTRS{idVendor}=="046d", ATTRS{idProduct}=="c24d", TAG+="uaccess"

KERNEL=="uinput", TAG+="uaccess"
```
I testing on some other keyboards, too. So they are also included here.

I won't merge any of this into master if I not got a full working concept, so don't worry things getting messed up. I read the libusb documentation as well as the one from uinput and python-uinput. That's the recommended way of accessing it. Since we run on root privileges right now we didn't need to take action to access it. 

There are mainly two ways to access:
`uaccess` meas user access there is another option `gaccess`, which will allow group access
This will set the owner of the stream to root:users where users seams to be a virtual group or something.

The other option is to gain access with the file permissions on the stream.You can set `MODE` to set them. This method is not recommended since you need to set 666 to make it universal and this is a big security problem.

All I am doing now is under the assumption the driver can be installed on a user level and is handled by systemd in that way, so we can fix the problems triggering commands not beeing logged in or not even on a tty. That's the reason why It's in the title.

My idea right now is to allow a install on user level, because there is maybe someone not allowed for sudo wanting to use the driver and can't right now. Best would be to detect root privileges on install and just install user level service then. I have to read some more on this, but I think systemd will handle start/stop of the service if we install it with the right target.

**Comment by braoult** [comment]
> My idea right now is to allow a install on user level

This is exactly the point I raised. I am unsure on how it will be handled when 2 users are logged-in on same `seat` (same keyboard etc.) at the same time, one only being active, and the user switch happening when using a `control-alt-F?` combination.

Note: During a switch, there is no login (the user is already there, the active window is still the one which was active during last Bob' switch-out to Alice).

My question was simply : What will be the owner and access right of your device after we switch from Alice to Bob ? And what will happen to Alice and Bob systemd user services (there are two of them, right ?) during a user switch ?

I believe the system is designed to make proper switch (for example X is grabbing/releasing devices, if I understand, but I am not sure of that).

In other words, I wonder if our "driver" will also have to do something (when we switch in/to or out/from a user), if the two services will be notified of some event, or if everything will be auto-magically handled.

**Comment by suabo** [comment]
>In other words, I wonder if our "driver" will also have to do something (when we switch in/to or out/from a user), if the two services will be notified of some event, or if everything will be auto-magically handled.

If there is no option in the service unit then there will be probably a systemwide signal we can listen to on a user switch. Just assuming... I will read into on demand.

I feel like the discussion is more related to "Install User-Level-Service" and maybe #63, then to the implementation of pyhidapi, which wasn't intended. Thats why I'll remove the "Install User-Level-Service" part from the headline now.  
If you interested in "installing without root" as a feature feel free to open a feature request and refer this issue.

**Comment by braoult** [comment]
> I feel like the discussion is more related to "Install User-Level-Service"

It is in indeed. The current `root service` works as expected with [this solution](https://github.com/JSubelj/g910-gkey-macro-support/blob/master/docs/Configuration-strict-user-environment.org), as there is no driver ownership by users (applications will simply get an input generated by root user).

I think that, if some device get user ownership and permission, this discussion is very important (in particular when two users want to run the same service).

But if you separate the root ownership from the `pyhidapi` migration, I agree it (this discussion) could be done somewhere else, when the multi-user permissions will be a subject.

**Comment by Luk164** [comment]
Hey, did anyone ever manage to the this done so OpenRGB can work with the driver enabled?

**Comment by suabo** [comment]
I'm currently working on the topic. Basically it's working right now. But I haven't tested anything so there will pass some time till I am going to release it. I'm going to push it later into the development branch. Maybe you want to test it.

**Comment by Luk164** [comment]
@suabo Sure, lmk and I will do some testing. I already took a crack at it myself but was unsuccessful, no matter what I did I couldn't make OpenRGB see the kb and for the driver to work at the same time

**Comment by suabo** [comment]
You can check the Version in the [dev-0.5.x](https://github.com/JSubelj/g910-gkey-macro-support/tree/dev-0.5.x) branch. It should work together with OpenRGB now, since the driver doesn't need to detach the usb interface vom the kernel.
I am running it with python 3.12 on ubuntu 24.04. Please report back if this works for you or if there are any errors.

**Comment by Luk164** [comment]
@suabo Took me a while and the results were not great. I am running CachyOS and used pipx to setup the venv with these packages:

```
> pipx runpip g910-gkeys list
DEBUG uv 0.10.10
DEBUG Checking for Python interpreter at path `bin/python`
Using Python 3.12.13 environment at: .
Package       Version
------------- -------
g910-gkeys    0.5.0
hidapi        0.15.0
python-uinput 1.0.1
```

With the driver running:

<img width="903" height="551" alt="Image" src="https://github.com/user-attachments/assets/3b4bd937-6408-46c2-9477-9e69e5ec326f" />

Once I stop it:

<img width="900" height="566" alt="Image" src="https://github.com/user-attachments/assets/592a17b6-b2c2-4037-bd7a-14b3b7d55a07" />

Double checking I am running the right version:
```
> g910-gkeys --version
g910-gkeys 0.5.0 by Jan Šubelj
```

Logs:
```
> g910-gkeys
2026-05-13 19:41:15,643 [INFO] {g910_gkeys.g910_gkeys} - --------------------------------------------------------------------------------
2026-05-13 19:41:15,643 [INFO] {g910_gkeys.g910_gkeys} - ----------------------STARTED g910-keys-pid:165447------------------------------
2026-05-13 19:41:15,643 [INFO] {g910_gkeys.g910_gkeys} - --------------------------------------------------------------------------------
2026-05-13 19:41:15,695 [INFO] {g910_gkeys.lib.hid_device} - Keyboard Logitech G910 Orion Spectrum found.
2026-05-13 19:41:15,700 [INFO] {g910_gkeys.misc.config} - Reading config from /home/luk/.config/g910-gkeys/config.json
2026-05-13 19:41:16,200 [INFO] {g910_gkeys.lib.hid_device} - Trying to disable f-key to g-key binding
2026-05-13 19:41:16,206 [INFO] {g910_gkeys.lib.hid_device} - G-key-mode - Disabled successfully
2026-05-13 19:41:17,411 [INFO] {g910_gkeys.g910_gkeys} - Change profile to MEMORY_1.
2026-05-13 19:41:21,312 [INFO] {g910_gkeys.g910_gkeys} - Change profile to MEMORY_2.
2026-05-13 19:41:21,957 [INFO] {g910_gkeys.g910_gkeys} - Change profile to MEMORY_1.
```

OpenRGB:
```bash
> openrgb --version
OpenRGB 0.9+ (1.0rc2), for controlling RGB lighting.
  Version:               0.9+ (1.0rc2)
  Build Date             Mon, 30 Mar 2026 14:06:54 +0000
  Git Commit ID          
  Git Commit Date        
  Git Branch 
```

**Comment by suabo** [comment]
@Luk164 I have some good news on this front. 

I tried myself with the newest version of openRGB and can confirm the problem. I figured that using hidapi with `import hid` still uses libusb in the background which takes full control over the opened interface. Replacing it with `import hidraw` did the trick. Now it is really using hidraw which allows the script, openRGB and other software to share the device for I/O.

I pushed the changes in the dev-0.5.x branch and bumped the version to 0.5.1.
I would appreciate if you can test it again and report back if it now works for you.

**Comment by Luk164** [comment]
@suabo It worked! Now I can have RGB controls and macros!
